### PR TITLE
[http_request] Fix ESP-IDF follow redirect

### PR DESCRIPTION
**Pull request by dentra**
# What does this implement/fix?

Follow redirects ESP-IDF http client implementation works only when using esp_http_client_perform, so current implementation completely does not work. We can't use esp_http_client_perform because of stream reading (esp_http_client_read), so I implement redirect functionality internally.

~~I also added parameters to change the client's read/write buffer, this will allow, for example, to process redirects from Github correctly.~~ 

At this moment request and response are limited to 512 bytes. PR #7125 resolve this limitation.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
http_request:
  buffer_size_rx: 1024
  buffer_size_tx: 1024
ota:
  - platform: http_request
    id: ota_http_request
update:
  - platform: http_request
    name: Firmware
    device_class: firmware
    source: https://github.com/$OWNER/$REPO/releases/latest/download/manifest.json
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/http_request/__init__.py [hunk 1]
```diff
@@ -110,6 +112,8 @@ def _declare_request_class(value):
                 cv.positive_not_null_time_period,
                 cv.positive_time_period_milliseconds,
             ),
+            cv.Optional(CONF_BUFFER_SIZE_RX): cv.All(cv.uint16_t, cv.only_with_esp_idf),
+            cv.Optional(CONF_BUFFER_SIZE_TX): cv.All(cv.uint16_t, cv.only_with_esp_idf),
```
Comment (jesserockz) [inline esphome/components/http_request/__init__.py]: The default is 512 for both of these in esp-idf, so we may as well set that here using `cv.SplitDefault`

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/http_request/__init__.py (hunk 1 above)
Comment (jesserockz) [reply esphome/components/http_request/__init__.py]: Also, as the buffer size is a new feature, it wont be merged into a patch release, but the redirect can be if you split the buffer size setting into its own PR.

Comment (dentra) [reply esphome/components/http_request/__init__.py]: > The default is 512 for both of these in esp-idf, so we may as well set that here using `cv.SplitDefault`

Done

> Also, as the buffer size is a new feature, it wont be merged into a patch release,

If the workaround using external_components is not suitable, of course I will do it.

Comment (dentra) [reply esphome/components/http_request/__init__.py]: Mmm, it would be necessary to update the documentation...

Comment (jesserockz) [reply esphome/components/http_request/__init__.py]: > If the workaround using external_components is not suitable, of course I will do it.

Well I would like to be able to release the redirect fixes in a patch release, but cant do that if it is included in the same PR as a new feature.
If they are in their own PRs then I can =)

Comment (dentra) [reply esphome/components/http_request/__init__.py]: Now split buffer size configuration to PR #7125

Comment (dentra) [reply esphome/components/http_request/__init__.py]: But two PRs broke of external_components usage 😢

Comment (jesserockz) [reply esphome/components/http_request/__init__.py]: If you want to combine them for use externally, then you just push a new branch on your repo and dont open a PR and use that branch in yaml =)

**Review by jesserockz — APPROVED** [review]

**Pull request merged**